[cling] Move from legacy to new pass manager

devajithvs opened the pull request:
# This Pull request:
- Moves away from the legacy pass manager to the new llvm pass manager. Care is take to ensure the same behavior before and after the update. Could benefit from more testing.

## Changes or fixes:
- Adapt the existing passes for the new pass manager.

## Checklist:

- [x] tested changes locally
- [ ] updated the docs (if necessary)

This PR fixes #

devajithvs commented:
- It would also be good to verify if the existing passes in BackendPasses.cpp are still required. But I think that's better done in a different PR.

devajithvs commented:
As an initial test, running tests in tutorials/dataframe returns the similar CPU timings:

Current master branch:
![master](https://github.com/root-project/root/assets/29475282/f38dd84e-218c-4a26-825c-a6f026ce384d)

With updated pass manager:
![newpassmanager](https://github.com/root-project/root/assets/29475282/1ba6421e-1880-4996-9251-b13f66080d77)

devajithvs commented:
WIP: Checking and fixing tests!

pcanal commented:
< Use new instance of all the pass/analysis managers every time 

If this commit works out, please enhance the commit log with an explanation of why this change is improving things.

hahnjo commented:
As a side remark, it sounds expensive to recreate all managers and passes for every module. Is this also what upstream does if it runs on multiple input files?

devajithvs commented:
> As a side remark, it sounds expensive to recreate all managers and passes for every module. Is this also what upstream does if it runs on multiple input files?

I agree that it might not be very efficient to recreate all managers/passes for every module. I've been looking upstream and couldn't find any instance where populating pass/analysis managers and running passes are separated unlike the legacy pass manager. I used this as a reference for this PR: https://reviews.llvm.org/D123425

[This issue](https://discourse.llvm.org/t/segmentation-faults-running-the-new-llvm-modulepassmanager-with-default-pipeline/59105) is what I think happened with the failing tests for us. From the thread, they recommend using a new instance every time we perform codegen as the analyses might not be cleared.

I could also try using the clear() method instead,

hahnjo commented:
Indeed, I had a quick look and not even (Thin)LTO reuses the `ModulePassManager` and its passes. So let's go with the most straight-forward solution of creating new passes on every run, the same what Clang does. If that proves to be a problem, we can still think how to optimize...

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -16,9 +16,10 @@
 #include "llvm/IR/Module.h"
 #include "llvm/IR/Verifier.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
```
Do we need `llvm/Passes/PassPlugin.h`? Instead, I think we should definitely include `llvm/IR/PassManager.h` because it defines things such as `PassInfoMixin`.

devajithvs replied:
Thanks for the review. 
I added that to try loading the SYCL passes in cling. Removed!

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -382,56 +353,44 @@ void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel)
   if (m_CGOpts.VerifyModule)
-      m_FPM[OptLevel]->add(createVerifierPass());
-  PMBuilder.populateFunctionPassManager(*m_FPM[OptLevel]);
+    m_MPM.addPass(VerifierPass());
```
The verifier pass should be earlier, at the very least before the default passes

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -344,36 +339,12 @@ BackendPasses::~BackendPasses() {
-#endif
-
+void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel,
+                                 llvm::ModulePassManager& m_MPM,
```
please rename to
```suggestion
                                 llvm::ModulePassManager& MPM,
```
the `m_` prefix is for member variables.

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -344,36 +339,12 @@ BackendPasses::~BackendPasses() {
-    Inlining = CodeGenOptions::NormalInlining;
-#endif
-
+void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel,
```
I believe we don't need the `llvm::Module` anymore, do we?

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -441,8 +400,16 @@ void BackendPasses::runOnModule(Module& M, int OptLevel) {
 
-  if (!m_MPM[OptLevel])
-    CreatePasses(M, OptLevel);
+  ModulePassManager m_MPM;
```
please rename to
```suggestion
  ModulePassManager MPM;
```

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.h:
```diff
@@ -10,7 +10,8 @@
 #define CLING_BACKENDPASSES_H
 
-#include "llvm/IR/LegacyPassManager.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Passes/PassPlugin.h"
```
Do we need anything from these headers? Instead, I believe we need the following:
```c++
#include "llvm/Analysis/CGSCCPassManager.h"
#include "llvm/Analysis/LoopAnalysisManager.h"
#include "llvm/Passes/PassManager.h"
```

devajithvs commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -382,56 +353,44 @@ void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel)
 
-  // Set up the per-module pass manager.
-  m_MPM[OptLevel].reset(new legacy::PassManager());
+    m_MPM = PB.buildPerModuleDefaultPipeline(Level);
```
Should we try to follow what's [upstream](https://github.com/root-project/root/blob/master/interpreter/llvm-project/clang/lib/CodeGen/BackendUtil.cpp#L763) or the existing logic?

From what I understand, BackendPasses.cpp had the same logic as BackendUtil.cpp except running a few passes of our own before running the default llvm passes. But they have diverged a lot.

hahnjo replied:
Probably we want to get closer to upstream logic, but the decision to run the always-inliner seems pretty conscious...

hahnjo commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -366,9 +373,28 @@ void BackendPasses::CreatePasses(llvm::Module& M, int OptLevel,
+
+    // Register a callback for disabling all other inliner passes.
+    PIC.registerShouldRunOptionalPassCallback([](StringRef P, Any) {
+      if (P.contains("InlinerPass"))
+        return false;
```
Very likely, this condition is too broad matching all `InlinerPass`es. But please confirm locally that future changes work on platforms where they didn't before, we should not waste CI resources for this...

devajithvs commented:
Result after grepping "inline" for all the registered passes (for -O1):
```
Module passes:
  always-inline
  inliner-wrapper
  inliner-ml-advisor-release
  print<inline-advisor>
  inliner-wrapper-no-mandatory-first
  partial-inliner
  scc-oz-module-inliner
  module-inline

Module analyses:
  inline-advisor

CGSCC passes with params:
  inline<only-mandatory>

Function passes:
  partially-inline-libcalls
  print<inline-cost>
  print<inliner-size-estimator>

Function passes with Params:
  ee-instrument<post-inline>

Function analyses:
  inliner-size-estimator
```

Corresponding passes:

```
Module passes:
  always-inline                        - AlwaysInlinerPass
  inliner-wrapper                      - ModuleInlinerWrapperPass
  inliner-ml-advisor-release           - ModuleInlinerWrapperPass
  print<inline-advisor>                - InlineAdvisorAnalysisPrinterPass
  inliner-wrapper-no-mandatory-first   - ModuleInlinerWrapperPass
  partial-inliner                      - PartialInlinerPass
  scc-oz-module-inliner                - buildInlinerPipeline
  module-inline                        - ModuleInlinerPass
Module analyses:
  inline-advisor                       - InlineAdvisorAnalysis
CGSCC passes with params:
  inline<only-mandatory>               - InlinerPass
Function passes:
  partially-inline-libcalls            - PartiallyInlineLibCallsPass
  print<inline-cost>                   - InlineCostAnnotationPrinterPass
  print<inliner-size-estimator>        - InlineSizeEstimatorAnalysisPrinterPass
Function passes with Params:
  ee-instrument<post-inline>           - EntryExitInstrumenterPass // Not disabled
Function analyses:
  inliner-size-estimator               - InlineSizeEstimatorAnalysis
```

Passes disabled with the callback:
```
AlwaysInlinerPass
ModuleInlinerWrapperPass
InlineAdvisorAnalysisPrinterPass
PartialInlinerPass
buildInlinerPipeline
ModuleInlinerPass
InlinerPass
```

devajithvs commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -344,94 +344,75 @@ BackendPasses::~BackendPasses() {
+
+    // Register a callback for disabling all other inliner passes.
+    PIC.registerShouldRunOptionalPassCallback([](StringRef P, Any) {
+      if (P.equals("AlwaysInlinerPass") ||
```
Oops. Fixed it. I'll also verify that the always inliner is running and report back.

devajithvs replied:
TODO: It'd be more efficient to use a set of passes to exclude somehow.

devajithvs commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -344,94 +344,75 @@ BackendPasses::~BackendPasses() {
+    OptimizationLevel Level = mapToLevel(m_CGOpts);
+    if (m_CGOpts.OptimizationLevel == 0) {
+      // TODO: Remove this after https://reviews.llvm.org/D146200
+      MPM = PB.buildO0DefaultPipeline(Level);
```
My bad. Thanks

devajithvs commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -16,13 +16,17 @@
 #include "llvm/Transforms/IPO/AlwaysInliner.h"
 #include "llvm/Transforms/IPO/Inliner.h"
+#include "llvm/Transforms/IPO/ModuleInliner.h"
+#include "llvm/Transforms/IPO/PartialInlining.h"
```
Removed

devajithvs commented on interpreter/cling/lib/Interpreter/BackendPasses.cpp:
```diff
@@ -441,8 +422,19 @@ void BackendPasses::runOnModule(Module& M, int OptLevel) {
+
+  // With the new pass manager, module analysis (MAM) can get modified when
+  // passes are run. It has to be reinitialized for each run.
+  // if (MPM.isEmpty())
```
Done!

hahnjo commented:
@phsft-bot build with flags -DCTEST_TEST_EXCLUDE_NONE=ON

hahnjo commented:
This seems to work now and the changes LGTM in principle. We'll break running with `-O0` until the last commit, but I understand that's to make it easier to revert later on. The message of the first commit should be reflowed to not be one very long line.

Otherwise, is this ready @devajithvs ? ping @vgvassilev

devajithvs commented:
It is ready. Unless there are more comments.

dpiparo reviewed (approved):
Good job @devajithvs . I am in favour of merging this PR but before doing so could you please check that we do not experience performance regressions of any sort, for example:
- max RSS and walltime to boot root `root -b -q`
- max RSS and walltime to minimal root use `root -b -q -e "ROOT::RDataFrame(8)"`
- max RSS and walltime to boot PyROOT `python -c "import ROOT"
- max RSS and walltime for minimal PyROOT use `python -c "import ROOT; ROOT.RDataFrame(8)"

Cheers,
     D

devajithvs commented:
> Good job @devajithvs . I am in favour of merging this PR but before doing so could you please check that we do not experience performance regressions of any sort, for example:
> 
>     * max RSS and walltime to boot root `root -b -q`
> 
>     * max RSS and walltime to minimal root use `root -b -q -e "ROOT::RDataFrame(8)"`
> 
>     * max RSS and walltime to boot PyROOT `python -c "import ROOT"
> 
>     * max RSS and walltime for minimal PyROOT use `python -c "import ROOT; ROOT.RDataFrame(8)"
> 
> 
> Cheers, D


Current master branch:
```bash
$ /usr/bin/ti/usr/bin/time -v root/builddir/bin/root -b -q
  ...
	Command being timed: "root/builddir/bin/root -b -q"
	User time (seconds): 0.08
	System time (seconds): 0.07
	Percent of CPU this job got: 98%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.16
	...
	Maximum resident set size (kbytes): 205412
    ...
```

```bash
$ /usr/bin/time -v root/builddir/bin/root -b -q -e "ROOT::RDataFrame(8)"
  ...
(ROOT::RDataFrame) An empty data frame that will create 8 entries
  ...
	Command being timed: "root/builddir/bin/root -b -q -e ROOT::RDataFrame(8)"
	User time (seconds): 0.23
	System time (seconds): 0.13
	Percent of CPU this job got: 98%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.38
	...
	Maximum resident set size (kbytes): 298432
    ...
```

With new passmanager:

```bash
/usr/bin/time -v root/newpassmanager/bin/root -b -q
  ...
	Command being timed: "root/newpassmanager/bin/root -b -q"
	User time (seconds): 0.08
	System time (seconds): 0.09
	Percent of CPU this job got: 94%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.18
	...
	Maximum resident set size (kbytes): 204860
	...
```


```bash
/usr/bin/time -v root/newpassmanager/bin/root -b -q -e "ROOT::RDataFrame(8)"
  ...
(ROOT::RDataFrame) An empty data frame that will create 8 entries
  ...
	Command being timed: "root/newpassmanager/bin/root -b -q -e ROOT::RDataFrame(8)"
	User time (seconds): 0.21
	System time (seconds): 0.16
	Percent of CPU this job got: 92%
	Elapsed (wall clock) time (h:mm:ss or m:ss): 0:00.41
	...
	Maximum resident set size (kbytes): 298312
	...
```

I have some issues trying to get pyroot working on my machine. Will edit this message after I get it working.

hahnjo reviewed (approved):
Looks good, let's merge and see if there is any fallout. If possible, please reflow the first commit message to wrap at 72 characters.

vgvassilev commented:
Could you post performance numbers before and after that change? I am interested in how much that affects single line statements as well as a 1000 line code snippet.

vgvassilev commented:
> > Could you post performance numbers before and after that change? I am interested in how much that affects single line statements as well as a 1000 line code snippet.
> 
> Did you see the numbers posted initially and just above last week? I think they cover the most relevant cases, including an instantiation of a a full `ROOT::RDataFrame`. And all examples where Clings is spending most of the time in the passes themselves don't really change, it's just how they are called...

Ok, overlooked that. If we are testing the pass manager re-creation over a every transaction, why we test things that do not have multiple transactions. That is, we need tests that call ProcessLine many many times to be able to measure the eventual overhead...

hahnjo commented:
Correct, but I argue that the current "real world" tests don't show any measurable overhead. If we find problems in micro-benchmarks, these can be addressed in followup changes. My rationale is that we *have* to go this step to the new pass manager, better make progress sooner than later...

dpiparo commented:
a run to `stressInterpreter` will not be a huge endeavour after all. It can be useful if in the end it turns out that 6.32 is the release that CMS and Alice pick up for the 2025 data taking.

devajithvs commented:
I'm not sure if this is the one (stressInterpreter - ctest):

Current Master:
```bash
dvalapar@emmanouil:~/work/root/builddir (master)$ ctest -R tectest -R test-stressinterpreter* --verbose
UpdateCTestConfiguration  from :/home/dvalapar/work/root/builddir/DartConfiguration.tcl
Parse Config file:/home/dvalapar/work/root/builddir/DartConfiguration.tcl
 Add coverage exclude regular expressions.
UpdateCTestConfiguration  from :/home/dvalapar/work/root/builddir/DartConfiguration.tcl
Parse Config file:/home/dvalapar/work/root/builddir/DartConfiguration.tcl
Test project /home/dvalapar/work/root/builddir
Constructing a list of tests
Ignore test: test-stressproof
Ignore test: roottest-cling-parsing-semicolon
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 338
    Start 338: test-stressinterpreter

338: Test command: /usr/bin/cmake "-DCMD=/home/dvalapar/work/root/builddir/test/stressInterpreter" "-DPRE=/usr/bin/cmake^-E^copy^/home/dvalapar/work/root/test/stressInterpreter.cxx^/home/dvalapar/work/root/builddir/test/stressInterpreter.cxx" "-DSYS=/home/dvalapar/work/root/builddir" "-P" "/home/dvalapar/work/root/builddir/RootTestDriver.cmake"
338: Environment variables: 
338:  ROOT_HIST=0
338: Test timeout computed to be: 1500
338: ****************************************************************************
338: *  Starting  stress INTERPRETER                                            *
338: ****************************************************************************
338: FuncCall ................................................................ OK
338: STLDict ................................................................. OK
338: Reflection .............................................................. OK
338: NestedStatements ........................................................ OK
338: ****************************************************************************
338: stressInterpreter: Real Time =   0.74 seconds Cpu Time =   0.74 seconds
338: ****************************************************************************
338: *  ROOTMARKS =12973.0   *  Root6.31/01   20231010/0
338: ****************************************************************************
338: (bool) false
1/1 Test #338: test-stressinterpreter ...........   Passed    1.71 sec

The following tests passed:
        test-stressinterpreter

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   1.97 sec
```

New pass manager:
```bash
dvalapar@emmanouil:~/work/root/newpassmanager (dev.newpassmanager)$ ctest -R test-stressinterpreter* --verbose
UpdateCTestConfiguration  from :/home/dvalapar/work/root/newpassmanager/DartConfiguration.tcl
Parse Config file:/home/dvalapar/work/root/newpassmanager/DartConfiguration.tcl
 Add coverage exclude regular expressions.
UpdateCTestConfiguration  from :/home/dvalapar/work/root/newpassmanager/DartConfiguration.tcl
Parse Config file:/home/dvalapar/work/root/newpassmanager/DartConfiguration.tcl
Test project /home/dvalapar/work/root/newpassmanager
Constructing a list of tests
Ignore test: test-stressproof
Ignore test: roottest-cling-parsing-semicolon
Done constructing a list of tests
Updating test list for fixtures
Added 0 tests to meet fixture requirements
Checking test dependency graph...
Checking test dependency graph end
test 338
    Start 338: test-stressinterpreter

338: Test command: /usr/bin/cmake "-DCMD=/home/dvalapar/work/root/newpassmanager/test/stressInterpreter" "-DPRE=/usr/bin/cmake^-E^copy^/home/dvalapar/work/root/test/stressInterpreter.cxx^/home/dvalapar/work/root/newpassmanager/test/stressInterpreter.cxx" "-DSYS=/home/dvalapar/work/root/newpassmanager" "-P" "/home/dvalapar/work/root/newpassmanager/RootTestDriver.cmake"
338: Environment variables: 
338:  ROOT_HIST=0
338: Test timeout computed to be: 1500
338: ****************************************************************************
338: *  Starting  stress INTERPRETER                                            *
338: ****************************************************************************
338: FuncCall ................................................................ OK
338: STLDict ................................................................. OK
338: Reflection .............................................................. OK
338: NestedStatements ........................................................ OK
338: ****************************************************************************
338: stressInterpreter: Real Time =   0.58 seconds Cpu Time =   0.58 seconds
338: ****************************************************************************
338: *  ROOTMARKS =16551.7   *  Root6.31/01   20231010/0
338: ****************************************************************************
338: (bool) false
1/1 Test #338: test-stressinterpreter ...........   Passed    1.51 sec

The following tests passed:
        test-stressinterpreter

100% tests passed, 0 tests failed out of 1

Total Test time (real) =   1.72 sec

```

vgvassilev commented:
Pretty nice! Thank you!

vgvassilev reviewed (approved):
LGTM!
